sound/soc/sof/intel/mtl.c

```diff
@@ -720,6 +721,70 @@ static void mtl_ipc_dump(struct snd_sof_dev *sdev)
 		hipcida, hipctdr, hipcctl);
 }
 
+/* 
+ * copied from hda_dsp_remove. Only difference is
+ * hda_dsp_core_reset_power_down() is replaced with mtl_dsp_core_power_down()
+ */
+int mtl_dsp_remove(struct snd_sof_dev *sdev)
+{
+	struct sof_intel_hda_dev *hda = sdev->pdata->hw_pdata;
+	const struct sof_intel_dsp_desc *chip = hda->desc;
+	struct hdac_bus *bus = sof_to_bus(sdev);
+	struct pci_dev *pci = to_pci_dev(sdev->dev);
+	struct nhlt_acpi_table *nhlt = hda->nhlt;
+
+	if (nhlt)
+		intel_nhlt_free(nhlt);
+
+	/* cancel any attempt for DSP D0I3 */
+	cancel_delayed_work_sync(&hda->d0i3_work);
+
+#if IS_ENABLED(CONFIG_SND_SOC_SOF_HDA)
+	/* codec removal, invoke bus_device_remove */
+	snd_hdac_ext_bus_device_remove(bus);
+#endif
+
+	hda_sdw_exit(sdev);
+
+	if (!IS_ERR_OR_NULL(hda->dmic_dev))
+		platform_device_unregister(hda->dmic_dev);
+
+	/* disable DSP IRQ */
+	snd_sof_dsp_update_bits(sdev, HDA_DSP_PP_BAR, SOF_HDA_REG_PP_PPCTL,
+				SOF_HDA_PPCTL_PIE, 0);
+
+	/* disable CIE and GIE interrupts */
+	snd_sof_dsp_update_bits(sdev, HDA_DSP_HDA_BAR, SOF_HDA_INTCTL,
+				SOF_HDA_INT_CTRL_EN | SOF_HDA_INT_GLOBAL_EN, 0);
+
+	/* disable cores */
+	if (chip)
+		mtl_dsp_core_power_down(sdev, chip->host_managed_cores_mask);
```

Comment:
we could explicitly use primary core mask as we know thats the only one needed for MTL to be powered down

Reply:
Wondering if we really need to copy this code, just one function that differs. We could have a callback instead and use the default code, as we did for interrupts

Reply:
@plbossart i thought about that too but we already have core_set_state() and the core_get/core_put. Adding another ops for powering down cores would just add to confusion. Not sure, if it makes sense or not

Reply:
core_get/core_put is a bit different then what we would need here as those two is also sending IPC message to the firmware and cure_put is not implemented for non TGL.
Here we do not want to send IPC.
`struct sof_intel_dsp_desc` level `power_down_host_managed_cores(struct snd_sof_dev *sdev)` or an even more generic `set_host_managed_cores_power(struct snd_sof_dev *sdev, bool power_up)` might work and can be used to share more code, but I don't have objection for the duplicated code.

@ranj063, @bardliao, should we implement core_get/put for MTL? I think we have all the ingredients, no?

Reply:
> @plbossart i thought about that too but we already have core_set_state() and the core_get/core_put. Adding another ops for powering down cores would just add to confusion. Not sure, if it makes sense or not

What's more confusing is the copy of all the initialization flow. That's not really maintainable IMHO, we should try to do something better where the flow remains the same except for one function.